File: 2023 KAKAO blind recuitment/1st/q6.py
# %%

# def solution(n, m, x, y, r, c, k):
#     queue = [(x, y, "")]
#     while len(queue) > 0:
#         [cx, cy, way] = queue.pop(0)
#         if (cx <= 0 or cx > n or cy <= 0 or cy > m):
#             continue
#         if(len(way) == k):
#             if(cx == r and cy == c):
#                 return way
#         else:
#             queue.append((cx + 1, cy, way + "d"))
#             queue.append((cx, cy - 1, way + "l"))
#             queue.append((cx, cy + 1, way + "r"))
#             queue.append((cx - 1, cy, way + "u"))
#     return "impossible"

import sys
sys.setrecursionlimit(2500)
# %%


def solution(n, m, x, y, r, c, k):
    def dfs(cx, cy, way):
        if (cx <= 0 or cx > n or cy <= 0 or cy > m):
            return None
        
        if(len(way) == k):
            if(cx == r and cy == c):
                return way
            return None
        #
        dx = abs(cx - r)
        dy = abs(cy - c)
        if((dx + dy) > (k - len(way))):
            return None
        #
        if (case := dfs(cx + 1, cy, way + "d")) != None:
            return case
        if (case := dfs(cx, cy - 1, way + "l")) != None:
            return case
        if (case := dfs(cx, cy + 1, way + "r")) != None:
            return case
        if (case := dfs(cx - 1, cy, way + "u")) != None:
            return case
        return None
    return v if (v := dfs(x, y, "")) != None else "impossible"


print(solution(50, 50, 3, 3, 3, 3, 2500))
# d l r u

# du
# lr
# rl
# ud

File: 2023 KAKAO blind recuitment/1st/q6.cpp
#include <string>
#include <vector>

using namespace std;

string dfs(int n, int m, int r, int c, int k, int cx, int cy, string way) {
    int dx = abs(cx - r);
    int dy = abs(cy - c);
    if ((dx + dy) > (k - way.length())) {
        return "impossible";
    }
    if (cx <= 0 || cx > n || cy <= 0 || cy > m) {
        return "impossible";
    }
    if (way.length() == k) {
        if (cx == r && cy == c) {
            return way;
        }
        return "impossible";
    }
    string wh;
    //
    wh = dfs(n, m, r, c, k, cx + 1, cy, way + "d");
    if (wh != "impossible") {
        return wh;
    }
    //
    wh = dfs(n, m, r, c, k, cx, cy - 1, way + "l");
    if (wh != "impossible") {
        return wh;
    }
    //
    wh = dfs(n, m, r, c, k, cx, cy + 1, way + "r");
    if (wh != "impossible") {
        return wh;
    }
    //
    wh = dfs(n, m, r, c, k, cx - 1, cy, way + "u");
    if (wh != "impossible") {
        return wh;
    }
    return "impossible";
}
string solution(int n, int m, int x, int y, int r, int c, int k) {
    int dx = abs(x - r);
    int dy = abs(y - c);
    int da = dx + dy;
    if((k - da) % 2 == 1){
        return "impossible";
    }
    return dfs(n, m, r, c, k, x, y, "");
}